File: 4_Oop/py_pkg/py_pkg/person_node.py
import rclpy
from rclpy.node import Node
import numpy as np

class Person_Node(Node):
    def __init__(self, node_name, 
                 PlannerName: str,
                 start: np.array,
                 goal: np.array):
        super().__init__(node_name)
        self.name = PlannerName
        self.start = start
        self.goal = goal 
        print("----------------")
        print(f"启用 {self.get_node_names()} 节点")
        print(f"包含 {self.name} 规划器")
        print("----------------")
    """
    判断是否抵达目标点
    """
    def isGoalReached(self, r: float) -> bool:
        # 计算欧氏距离
        distance = np.linalg.norm(self.start - self.goal)
        if distance < r:
            self.get_logger().info(f"距离小于 {r}")
            return True
        self.get_logger().info(f"距离大于 {r}")
        return False
    
def main():
    rclpy.init()
    node = Person_Node("P_Node","Teb",np.array([0.0, 1.0]), np.array([2.0, 2.0]))
    node.isGoalReached(1.0)
    rclpy.spin(node)
    rclpy.shutdown()

File: 4_Oop/cpp_pkg/src/person_node.cpp
#include <string>
#include "rclcpp/rclcpp.hpp"
#include <Eigen/Dense>

using namespace std;
using namespace Eigen;

// 公共继承节点类
class Person_Node : public rclcpp::Node
{

private:
    string node_name;
    string PlannerName;
    Array<float, 2, 1> start;
    Array<float, 2, 1> goal;

public:
    Person_Node(
        const string &_nodename,
        const string &_PlannerName,
        Array<float, 2, 1> &_start,
        Array<float, 2, 1> &_goal) : Node(_nodename),
                                     PlannerName(_PlannerName),
                                     start(_start),
                                     goal(_goal)
    {
        this->node_name = _nodename;
    };

    bool isGoalReached(double r)
    {
        Array<float, 2, 1> difference = this->goal - this->start;
        double distance = difference.matrix().norm();
        if (distance < r)
        {
            RCLCPP_INFO(this->get_logger(), "Goal reached! Distance: %f", distance);
            return true;
        }
        RCLCPP_INFO(this->get_logger(), "Goal not reached. Distance: %f", distance);
        return false;
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    Array<float, 2, 1> start(0, 0);
    Array<float, 2, 1> goal(5, 5);
    auto node = make_shared<Person_Node>("person_node", "AStarPlanner", start, goal);
    node->isGoalReached(1.0);
    rclcpp::spin(node);
    rclcpp::shutdown();
}